core/src/main/java/io/parsingdata/metal/data/ParseValueCache.java

```diff
@@ -0,0 +1,85 @@
+package io.parsingdata.metal.data;
+
+import static io.parsingdata.metal.data.Selection.NO_LIMIT;
+import static io.parsingdata.metal.data.Selection.reverse;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.token.Token;
+
+public class ParseValueCache {
+
+    public static final ParseValueCache NO_CACHE = new ParseValueCache(null);
+
+    private final Map<String, ImmutableList<ParseValue>> cache;
+
+    /**
+     * Start a cache that keeps track of values added to the parse graph.
+     * <p>
+     * In case no caching is desired, {@link #NO_CACHE} should be used instead.
+     */
+    public ParseValueCache() {
+        this(new HashMap<>());
+    }
+
+    // For internal use only. It is private to avoid setting the cache to null. The NO_CACHE constant should be used instead.
+    private ParseValueCache(final Map<String, ImmutableList<ParseValue>> cache) {
+        this.cache = cache;
+    }
+
+    public Optional<ImmutableList<Value>> find(final String scopeName, int limit) {
+        if (this == NO_CACHE) {
+            return Optional.empty();
+        }
+        final String s = shortName(scopeName);
+        ImmutableList<ParseValue> valueImmutableList = cache.getOrDefault(s, new ImmutableList<>());
+        ImmutableList<Value> result = new ImmutableList<>();
```

Comment:
With these two mutable variables we are letting go of the mostly functional style Metal was written in. I don't mind necessarily, it's just that we should be aware that we're changing our style.

Reply:
Yes, good point. For consistency changed it to use trampolines.